[IV-23-24] Objetivo 5

raultl12 opened the pull request:
* [x] ¿Se han establecido criterios previos de elección de la imagen base, así
como los criterios de búsqueda que han conducido a las imágenes que se están
considerando?
* [x] ¿Se ha tenido en cuenta que la imagen es para testear, y que por tanto
tiene una función específica, tal como testear cómo va nuestra aplicación en la
última versión publicada del lenguage?
* [x] ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?
* [x] ¿Se han tenido en cuenta criterios reales para la elección, y no criterios
irrelevantes como cuanto tiempo se tarda en construirla? (Se va a construir una
vez, se va a usar cientos o miles de veces)
* [x] ¿Se han seguido las buenas prácticas en las órdenes que construyen
   la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?
* [x] ¿La imagen incluye sólo y exclusivamente la infraestructura
   necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
   "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
   comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
   nuestro proyecto?
* [x] ¿Se han documentado y enlazado los commits a las imágenes que se
   han testeado?
* [x] ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los
   tests que se van a lanzar?
* [x] ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
en vez del directorio de trabajo?
* [x] ¿Se han establecido claramente los criterios de búsqueda y de aceptación de
   la imagen base?
* [x] ¿Tienes claro cuales son las condiciones en las que tiene que actualizarse
la imagen en DockerHub y has configurado el workflow de acuerdo con ello? En
particular, ¿sabes qué condiciones tienen que darse para que se actualice y qué
condiciones *no* deben darse?
* [x] Si se usan GitHub actions para su publicación, ¿se han considerado
cuidadosamene los criterios (ramas, paths) para que se lancen?
* [x] ¿Se ha comprobado que el ENTRYPOINT del contenedor no escriba nada en el
directorio que monta, ya que no tiene derecho de escritura?
* [x] ¿Se ha abierto un shell en el contenedor una vez construida la imagen y
comprobado que los ficheros que se necesitan están todos incluidos, y que no hay
ficheros que no se necesiten?

raultl12 committed:
add: se añade archivo deno lock

Se ha añadido el archivo deno lock porque no se le estaba haciendo seguimiento y es necesario para el entorno de pruebas

raultl12 committed:
doc: establecidos criterios de eleccion de imagen

Se han establecido los criterios de eleccion de la imagen base sobre la que montaremos el entorno para pruebas

raultl12 committed:
add: añadido dockerfile para contenedor closes #72

Se ha creado y añadido el archivo dockerfile que especifica las reglas para la construcción del contenedor de pruebas

raultl12 committed:
doc: añadida informacion de contenedor closes #74

Se ha añadido al readme informacion sobre el contenedor de pruebas. Se cierran tambien issues anteriores que por error no habian sido cerrados closes #71 closes #73

raultl12 committed:
add: github action para actualizar contenedor closes #76

Se ha creado una github action para que se actualice el contendor automaticamente cuando se modifican los archivos especificados

migueruiz commented on Dockerfile:
```diff
@@ -0,0 +1,25 @@
+
+COPY deno.json deno.lock velociraptor.ts ./
+
+RUN apt-get update
```
Deberías de cambiar la manera de escribir esto, en vez de escribir tantas veces "RUN" o "COPY", podrías hacerlo más corto y simple.

raultl12 replied:
Los copy no los puedo simplificar en uno solo ya que el destino de copia es distinto en cada uno. Aun así, los RUN si los he juntado en uno solo. Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/f37bfe2432ae09f97b300ce8ac8a16e95af8cfdd

migueruiz reviewed (changes requested):
Hola, seré tu revisor! Has hecho un buen trabajo, por ponerle algunas pegas, te dejo los comentarios de abajo. 😄

raultl12 committed:
refactor: simplificado comando RUN para que solo haya uno closes #77

Se han eliminado todas las apariciones del comando RUN y unido en un solo comando

raultl12 commented:
@migueruiz  he cambiado lo que me has dicho, revisalo cuando puedas

migueruiz reviewed (approved):
Todo correcto!

JJ commented:
¿Listo para revisar?

raultl12 commented:
@JJ Listo para revision

JJ reviewed (changes requested):
Los fuentes y tests que se tienen que usar deben montarse desde el directorio externo a la imagen **no** estar incluidos en la imagen. Del [guión](http://jj.github.io/IV/documentos/proyecto/5.Docker):

> ... hay que tener en cuenta es que las bibliotecas estarán dentro del contenedor, y los fuentes fuera del mismo, en un directorio donde no hay derechos de escritura.

raultl12 committed:
refactor: actualizado dockerfile closes #78

Se ha actualizado el dockerfile para que las fuentes y los test no se incluyan en la imagen sino que se monten desde el directorio externo

raultl12 commented:
@JJ listo para revisión

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+LABEL maintainer="raultlopez@correo.ugr.es"
+
+COPY deno.json deno.lock velociraptor.ts ./
```
¿Seguro que quieres copiar esto a un directorio que va a ser "pisado" por los fuentes?

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/3e61c54859705c61f3256e9ae27901434af0dbe8

JJ replied:
¿Seguro que necesitas tener `velociraptor.ts` *dentro* del contenedor? Cuanto ejecutes `vr`, usará el de `WORKDIR`, ¿no?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+COPY deno.json deno.lock velociraptor.ts ./
+
+RUN deno cache --lock=deno.lock --reload --lock-write deno.json
```
¿Con qué privilegios estás ejecutando esto?

raultl12 replied:
Se hace con privilegios root porque deno esta instalado en el usuario root. Así esta hecho en la imagen.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+RUN deno cache --lock=deno.lock --reload --lock-write deno.json
+
+RUN apt-get update && apt-get install unzip -y && deno upgrade && deno task install_vr && deno cache --unstable https://deno.land/std@0.207.0/assert/mod.ts
```
¿Estás actualizando `deno` después de ejecutar `deno cache`?

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/c7d690fef240a61f87996f26a758f4832444245d

JJ replied:
¿Dónde está codificada la tarea `install_vr`?

JJ replied:
Ah, OK, en deno.json; tiene sentido, claro.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM baleyko/deno:latest
+
+WORKDIR /app/test
```
De [errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-5.md):

> /app/test es sólo el directorio donde se va a "montar" (léase: pisar) el directorio donde se encuentra el repositorio; sólo es el directorio de trabajo para esa tarea. Cualquier cosa que se copie o se cree ahí no va ser visible en la imagen cuando se ejecute el ENTRYPOINT; por lo que es conveniente no hacer nada ahí. Para llevar a cabo instalaciones o cualquier otra cosa hay diferentes directorios que se pueden usar: $HOME o simplemente /app, que por eso tiene el directorio donde se monta dos componentes y no uno solo.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/3e61c54859705c61f3256e9ae27901434af0dbe8

JJ replied:
Lo habías resuelto, pero lo has vuelto a romper. Ahora estás copiando en un directorio que va a ser invisible cuando se monte el directorio externo, que era lo que habías arreglado al principio.
Por favor, no rompas cosas que ya habías arreglado tratando de arreglar cosas que no lo has hecho. `WORKDIR` no afecta a toda la imagen, afecta a las órdenes siguientes, o a `ENTRYPOINT` si se pone al final.

JJ commented:
>     * [x]  ¿Se han seguido las buenas prácticas en las órdenes que construyen
>       la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?


¿Es así realmente?

> 
>     * [x]  ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
>       en vez del directorio de trabajo?
> 

Por favor, comprueba que esto ha sido así.

JJ commented on docs/eleccion_imagen.md:
```diff
@@ -0,0 +1,15 @@
+
+* [denoland/deno](https://hub.docker.com/r/denoland/deno). Es la imagen oficial de deno para docker. 
+* [Baleyko/deno](https://hub.docker.com/r/baleyko/deno). Es una imagen no oficial que contiene ubuntu focal con deno ya instalado. 
+* [Ubuntu](https://hub.docker.com/_/ubuntu). Imagen oficial de Ubuntu. No trae nada instalado por lo que hay que instalar Deno y TypeScript.
```
Seguramente habrá alguna más, ¿no? ¿No hay otras imágenes de vendors oficiales o extraoficiales?

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/a3fe104ead54c55758143f42d6e3e19911b56ba9

JJ reviewed (changes requested):
Por favor, revisa lo indicado y darle una vuelta, que ya queda menos.

raultl12 committed:
fix: error de indentacion en github action closes #79

Se ha corregido el error de indentacion en el archivo

raultl12 committed:
fix: error en las ordenes de Dockerfile closes #80

Se ha cambiado el orden de los comandos en el dockerfile para que deno se actualice antes de ejecutar deno cache

raultl12 committed:
fix: mal workdir en dockerfile closes #82

Se ha cambiado el workdir de /app/test a /app

raultl12 committed:
doc: añadidas mas opciones de imagen closes #81

Se han añadido mas opciones de imagenes para deno ya que no habia suficientes imagenes

raultl12 commented:
@JJ Listo para revision

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+RUN apt-get update && apt-get install unzip -y && deno upgrade && deno task install_vr && deno cache --unstable https://deno.land/std@0.207.0/assert/mod.ts
+
+RUN deno cache --lock=deno.lock --reload --lock-write deno.json
```
¿Seguro que necesitas reescribir deno.lock con `--lock-write`?

raultl12 replied:
Diria que si, según esta pagina de [deno](https://docs.deno.com/runtime/manual/basics/modules/integrity_checking) se hace para que la version que corre en local y en una nueva maquina (en nuestro caso el docker) sean las mismas

JJ replied:
La versión ¿de qué?

JJ replied:
En esa página dice:

> --lock-write is used to output dependency hashes to the lock file

¿Para qué quieres hacerlo, si ya lo has generado y lo estás copiando desde fuera? ¿Has probado simplemente a quitarlo?

JJ replied:
No has resuelto este comentario.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/ec5b98860b66c68af0a8ab21f47b948f6a9a27e1

JJ commented on .github/workflows/contenedorTest.yml:
```diff
@@ -0,0 +1,27 @@
+    paths:
+      - 'Dockerfile'
+      - 'deno.json'
+      - 'deno.lock'
```
Estos paths son los ficheros que vas a meter en el contenedor, ¿Seguro que has puesto que se lance cuando cambien todos y cada uno de los ficheros?

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/f20408e4c2a3895ce64cf4788b5bd8dabeffe6e2

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+ENV PATH="/root/.deno/bin/bin:$PATH"
+
+ENTRYPOINT ["vr", "run", "test"]
```
¿En qué directorio se está ejecutando esto? ¿Seguro que se ejecuta en el que se ha especificado en el guión?

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/30882e734841c62e7d38f4d2d5da9431db9b31bd

JJ reviewed (changes requested):
Gracias por hacer los cambios, pero al hacerlos creo que has dejado algunas instrucciones sobre la imagen sin cumplir, como lo del directorio de trabajo en el que se tiene que ejecutar el ENTRYPOINT.
Por otro lado, los ficheros para el gestor de tareas van a estar en los fuentes. Que lo hayas tenido que copiar seguramente es consecuencia del problema de no haber establecido bien el `WORKDIR`. Pero si por alguna razón específica de `vr` hace falta, tendrás que tenerlo en cuenta cada vez que cambie para generar una nueva imagen.

raultl12 committed:
refactor: el dockerfile ya no copia velociraptor closes #83

El archivo dockerfile ya no copia velociraptor.ts. Simplemente no era necesario copiarlo

raultl12 committed:
refactor: cambiado directorio de ejecucion closes #84

El directorio de ejecucion que especifica el guion es /app/test, se ha cambiado a ese

raultl12 committed:
fix: error de sintaxis closes #85

Soluciona el error de sintaxis que aparecia en el archivo de github actions

raultl12 commented:
@JJ Listo para revision. Efectivamente no necesitaba copiar el archivo de velociraptor

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+WORKDIR /app/test
+
+LABEL maintainer="raultlopez@correo.ugr.es"
```
Las mejores prácticas también incluyen una etiqueta con la versión.

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/b22ba72a95a19bab353abde3c106bca5feaae6fd

JJ reviewed (changes requested):
Tienes todavía varios días para resolver este objetivo. Reflexiona un poco sobre lo que haces, tratando de no romper las cosas que ya habías arreglado, sobre todo, y mirando las cosas que sí has puesto, si realmente hacen falta o no.

raultl12 committed:
feat: añadida version al dockerfile closes #86

Se ha añadido la etiqueta con la version al dockerfile

raultl12 committed:
fix: corregido workdir closes #88

Se ha corregido el problema del workdir que habia roto en un commit anterior

raultl12 commented:
@JJ Listo para revision

JJ commented on docs/eleccion_imagen.md:
```diff
@@ -0,0 +1,17 @@
+
+* [denoland/deno](https://hub.docker.com/r/denoland/deno). Es la imagen oficial de deno para docker. 
+* [Baleyko/deno](https://hub.docker.com/r/baleyko/deno). Es una imagen no oficial que contiene ubuntu focal con deno ya instalado. 
+* [Ubuntu](https://hub.docker.com/_/ubuntu). Imagen oficial de Ubuntu. No trae nada instalado por lo que hay que instalar Deno y TypeScript.
```
¿Typescript por qué?

raultl12 replied:
Resuelto en: https://github.com/raultl12/TeamFinder/pull/75/commits/b3d826ec5dafd368d92493778c4c96f1732f4cee

JJ reviewed (changes requested):
Ya sabes que la regla es que hay que resolver *todos* los comentarios.

raultl12 committed:
fix: error en archivo eleccion imagen closes #89

Se ha corregido el error que indicaba que habia que instalar TypeScript con la imagen de ubuntu. Al usar Deno no hay que instalar TypeScript.

raultl12 committed:
refactor: eliminado lock-write de dockerfile closes #90

Se ha eliminado la opcion lock-write del comando deno cache porque no era necesaria al estar generando el archivo deno.json fuera de la imagen y simplemente copiandolo

raultl12 commented:
@JJ Listo para revision

JJ reviewed (approved):
:+1:
